mmcv/ops/diff_iou_rotated.py

```diff
@@ -0,0 +1,296 @@
+import torch
+from torch.autograd import Function
+
+from ..utils import ext_loader
+
+EPSILON = 1e-8
+ext_module = ext_loader.load_ext('_ext',
+                                 ['diff_iou_rotated_sort_vertices_forward'])
+
+
+class SortVertices(Function):
+
+    @staticmethod
+    def forward(ctx, vertices, mask, num_valid):
+        idx = ext_module.diff_iou_rotated_sort_vertices_forward(
+            vertices, mask, num_valid)
+        ctx.mark_non_differentiable(idx)
+        return idx
+
+    @staticmethod
+    def backward(ctx, gradout):
+        return ()
+
+
+def box_intersection(corners1, corners2):
+    """Find intersection points of rectangles.
+    Convention: if two edges are collinear, there is no intersection point.
+
+    Args:
+        corners1 (Tensor): (B, N, 4, 2) Corners of the first box.
+        corners2 (Tensor): (B, N, 4, 2) Corners of the second box.
+
+    Returns:
+        Tuple:
+         - Tensor: (B, N, 4, 4, 2) Intersections.
+         - Tensor: (B, N, 4, 4) Valid intersections mask.
+    """
+    # build edges from corners
+    # B, N, 4, 4: Batch, Box, edge, point
+    line1 = torch.cat([corners1, corners1[:, :, [1, 2, 3, 0], :]], dim=3)
+    line2 = torch.cat([corners2, corners2[:, :, [1, 2, 3, 0], :]], dim=3)
+    # duplicate data to pair each edges from the boxes
+    # (B, N, 4, 4) -> (B, N, 4, 4, 4) : Batch, Box, edge1, edge2, point
+    line1_ext = line1.unsqueeze(3)
+    line2_ext = line2.unsqueeze(2)
+    x1 = line1_ext[..., 0]
+    y1 = line1_ext[..., 1]
+    x2 = line1_ext[..., 2]
+    y2 = line1_ext[..., 3]
+    x3 = line2_ext[..., 0]
+    y3 = line2_ext[..., 1]
+    x4 = line2_ext[..., 2]
+    y4 = line2_ext[..., 3]
+    # math: https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection
+    num = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4)
+    den_t = (x1 - x3) * (y3 - y4) - (y1 - y3) * (x3 - x4)
+    t = den_t / num
+    t[num == .0] = -1.
+    mask_t = (t > 0) & (t < 1)  # intersection on line segment 1
+    den_u = (x1 - x2) * (y1 - y3) - (y1 - y2) * (x1 - x3)
+    u = -den_u / num
+    u[num == .0] = -1.
+    mask_u = (u > 0) & (u < 1)  # intersection on line segment 2
+    mask = mask_t * mask_u
+    # overwrite with EPSILON. otherwise numerically unstable
+    t = den_t / (num + EPSILON)
+    intersections = torch.stack([x1 + t * (x2 - x1), y1 + t * (y2 - y1)],
+                                dim=-1)
+    intersections = intersections * mask.float().unsqueeze(-1)
+    return intersections, mask
+
+
+def box1_in_box2(corners1, corners2):
+    """Check if corners of box1 lie in box2.
+    Convention: if a corner is exactly on the edge of the other box,
+    it's also a valid point.
+
+    Args:
+        corners1 (Tensor): (B, N, 4, 2) Corners of the first box.
+        corners2 (Tensor): (B, N, 4, 2) Corners of the second box.
+
+    Returns:
+        Tensor: (B, N, 4) Intersection.
+    """
+    a = corners2[:, :, 0:1, :]  # (B, N, 1, 2)
```

Comment:
a, b, d are not appropriate names, we usually do not use single-letter variable names.